src/App.jsx

```diff
@@ -1,27 +1,40 @@
 import TodoList from './TodoList';
 import AddTodoForm from './AddTodoForm';
-import { useState } from 'react';
+import React, { useState } from 'react';
+
+const useSemiPersistentState =()=>{
+  const [todoList ,setTodoList ] = useState(JSON.parse(localStorage.getItem('savedTodoList')) || []);
+  
+  React.useEffect(()=>{
+    localStorage.setItem('savedTodoList',JSON.stringify(todoList))
+  },[todoList]);
+  return [todoList ,setTodoList];
+};
 
 function App() {
+  const [todoList ,setTodoList ] = useSemiPersistentState();
 
-  const [newTodo, setnewTodo] = useState([]);
+  const addTodo =(newTodo) =>{
+    setTodoList([...todoList,newTodo]);
+  };
 
-  const SetNewTodo =(todoTitle) =>{
-    setnewTodo(todoTitle);   
+  const removeTodo =(id)=>{
+    const newTodoList = todoList.filter(
+      (todo)=> id !== todo.id
+    );
+    setTodoList(newTodoList);
   }
 
-   return (
-    <>
+  { /*used the Fragments <> </> inside the return function*/ }
+   return (  
+    <> 
       <div>
```

Comment:
You can simplify the code by removing the unnecessary <div> wrapper and using React fragments (<> </>) directly. Extra elements in the DOM can impact rendering performance, especially in large applications. By removing the unnecessary <div>, we can slightly improve performance, as there are fewer elements for the browser to process.

Reply:
"Aaand a stretch goal - how about finding a way to prevent a user from adding an empty todo title?"

I can use the IF condition statement to prevent that from happening, but I was just following the lesson. Maybe in the upcoming lesson, they will ask us to resolve I just don't want to rush